### DISCUSSION TITLE
Modal automatically opening on Wallet Disconnect

### DISCUSSION THREAD

**Discussion by nischitpra** [WalletConnect/walletconnect-monorepo#2480] [Developer Support] [answered]
When wallet is not connected and I open/reload the page, Wallet connect modal automatically opens. But when I do this after connecting my wallet, it works fine and doesn't open the modal automatically. As soon as i disconnect the wallet, the modal opens again. I have the most recent wagmi and web3modal versions.

```
    "@web3modal/ethereum": "^2.4.1",
    "@web3modal/react": "^2.4.1",
    "@web3modal/standalone": "^2.4.1",
    "wagmi": "^1.0.7",
    "viem": "^0.3.37",
```

The code I'm using is

```
import { configureChains, createConfig, WagmiConfig } from "wagmi";
import { mainnet, goerli, polygon, localhost, hardhat } from "wagmi/chains";
import { publicProvider } from "wagmi/providers/public";
import { jsonRpcProvider } from "wagmi/providers/jsonRpc";
import { EthereumClient } from "@web3modal/ethereum";
import { Web3Modal } from "@web3modal/react";
import "wagmi/window";

import { InjectedConnector } from "wagmi/connectors/injected";
import { WalletConnectConnector } from "@wagmi/core/connectors/walletConnect";
import { CoinbaseWalletConnector } from "@wagmi/core/connectors/coinbaseWallet";

window.Buffer = window.Buffer || Buffer;
window.global = window.global || window;
window.process = window.process || { env: {} }; // Minimal process polyfill


const { chains, publicClient, webSocketPublicClient } = configureChains(
  [mainnet, polygon, goerli, hardhat],
  [
    jsonRpcProvider({
      rpc: (chain) => ({
        http: supportedNetworks.apis.rpc[chain.id],
      }),
    }),
    publicProvider(),
  ]
);

const wagmiConfig = createConfig({
  autoConnect: true,
  connectors: [
    new InjectedConnector({ chains }),
    new WalletConnectConnector({
      chains,
      options: {
        projectId: WALLET_CONNECT_PROJECT_ID,
        showQrModal: true,
        qrModalOptions: {
          themeVariables: {
            "--w3m-z-index": 300,
          },
        },
      },
    }),
    new CoinbaseWalletConnector({
      chains,
      options: {
        appName: "Mercle",
      },
    }),
  ],
  publicClient,
  webSocketPublicClient,
});

const ethereumClient = new EthereumClient(wagmiConfig, chains);

const root = ReactDOM.createRoot(document.getElementById("root"));
root.render(
  <BrowserRouter>
    <WagmiConfig config={wagmiConfig}>
          <App />
    </WagmiConfig>
    <Web3Modal projectId={WALLET_CONNECT_PROJECT_ID} ethereumClient={ethereumClient} themeMode="dark" />
  </BrowserRouter>
);

```

The code I'm using to connect to a particular connector is

```
    const { connectAsync, connectors } = useConnect();
    const { open: openWeb3Modal } = useWeb3Modal();
  
     ...
 
     const walletConnect = async (connectorId, chainId = getChain()?.id) => {
         // if we are able to get value from wallet, then wagmi already has a connector
         if ((!connectorId || connectorId == getConnector()?.id) && getChain()) return;
         if (getConnector()) {
             disconnect();
         }

         const targetConnector = connectors.find((c) => c.id == connectorId);
         window.targetConnector = targetConnector;
         console.log(connectors, connectorId, targetConnector);
         if (!targetConnector) {
              openWeb3Modal();
              await waitForSigner(getSigner);
         } else {
              await connectAsync({ connector: targetConnector, chainId });
              await waitForSigner(getSigner);
          }
      };

    const waitForSigner = (getSigner) => {
          return new Promise((resolve, reject) => {
                const interval = setInterval(() => {
                    if (getSigner()) {
                        console.log("wallet connected");
                        clearInterval(interval);
                        resolve();
                    }
                 }, 200);
            });
      };
```

I found this other thread but its closed. Not sure if its related but its the closest I could find
https://github.com/rainbow-me/rainbowkit/issues/686#:~:text=If%20updating%20your%20React%20version%20does%20not%20help%2C,--save%20--save-exact%20wagmi%400.6.8%29%20and%20this%20fixed%20the%20issue.

**Comment by glitch-txs** [comment]
Hi, seems like you're mixing some dependencies. Do you want to use the web3modal with extension wallets like MetaMask or the standalone version only?

**Reply by nischitpra** [reply]
Hey @glitch-txs , I'm not exactly sure which library to use. Heres the usecase, I have a custom login panel that has option to connect with injected wallet, walletconnect and coinbase. For each I'm passing the connector to the `connectAsync` as mentioned above. So basically, when the user clicks on injected wallet, `InjectedConnector` is used and when clicking on wallet connect button,  `WalletConnectConnector` is used. So, the web3modal qr code should only appear when clicking on the button. 

Also, I tried downgrading wagmi to `1.0.5` and this issue doesn't appear. I've raised a discussion on wagmi as well (https://github.com/wagmi-dev/wagmi/issues/2440) but they said to contact with web3modal

**Comment by glitch-txs** [comment]
hi, I created a simplified example for this :) I only needed to install wagmi and viem, don't forget to add the project ID, tested and worked fine for injected, coinbase and WalletConnect. Wagmi uses WalletConnect standalone, if you want the full web3modal version you'll need to do some extra config
https://github.com/glitch-txs/wagmi-walletconnect

**Reply by nischitpra** [reply]
Hey thanks removing the extra `<Web3Modal/>` worked :+1: , there was another wagmi `1.0.8` update as well. Thanks :+1:

**Reply by nischitpra** [reply]
Hey @glitch-txs , How do I initiate connection to multiple chains? I have to perform network switch which was pretty straight forward in V1. But in V2, we have preconnect to the chains. Can't find a proper documentation on how to do this. I was hoping that wagmi's `configureChains` would have handled this internally but it doesnt.
I found these docs which just sets the default chain before connection. But I'm only able to set one so only the most recent chain set is connected. Also I'm not sure how to call these functions via wagmi. There are hooks for `@web3modal/react` but this doesn't work with `WalletConnectConnector` as mentioned above, they're different libraries.

https://docs.walletconnect.com/2.0/web3modal/html-js/actions#web3modalsetdefaultchain 
https://docs.walletconnect.com/2.0/web/providers/universal#multi-chain

**Reply by glitch-txs** [reply]
Hi, for web3modal yes you can use the hook `useWeb3modal` to get `setDefaultChain`

On the case of wagmi connector, wagmi utilizes the ethereum-provider package, for that one you should be fine by calling [useSwitchNetwork hook](https://wagmi.sh/react/hooks/useSwitchNetwork) from wagmi.

Keep in mind you should initialize your connection with all the chains at the same time, when you "switch"; the change will happen internally in the dapp but you won't get a request on the wallet (since the wallet already accepted the chain when connecting), to make sure this work you can trigger a transaction and check to which chain it is aimed for

For example:
1. You connect to mainnet and polygon (default is the first chain you choose).
2. You trigger a tx and it will aim mainnet
3. You request the switch to polygon (nothing externally happens in terms of UI)
4. You trigger a tx again and it aims to polygon now

**Reply by nischitpra** [reply]
@glitch-txs I'm using `useSwitchNetwork` to switch networks but its not working. The initial connection request only contains the first network the chains list. And since other networks are not approved (as it was not present in the initial connection), when I try and switch network I get the error
```
SwitchChainError: An error occurred when attempting to switch chain.

Details: Chain 'eip155:137' not approved. Please use one of the following: eip155:1
```

I tried adding polygon to the example code you've provided as well and its the same issue, the initial connection request only has ethereum in the networks because thats the first one in the list.

![Screenshot 2023-05-29 at 4 57 43 PM](https://github.com/WalletConnect/walletconnect-monorepo/assets/20742197/52d00354-43f9-40cd-9d6d-ca13bd1c31ba)

**Reply by glitch-txs** [reply]
useSwitchNetwork works fine for me, also by getting the current network it will show you the default one so it's not necesary to trigger the tx to check :)

I set the chains in the config first as well:
https://github.com/glitch-txs/wagmi-walletconnect/blob/main/src/pages/_app.tsx#L11

what connector are you using? also which wallet is that one?

**Reply by nischitpra** [reply]
Hey @glitch-txs 
Heres the exact same thing I'm doing.
I've updated `Connect.tsx` to have changeNetwork button. Its giving me the same error sadly.
https://github.com/nischitpra/wagmiWalletConnect/blob/c6aa11a8450ceafeba397430a2f5368862b665f4/src/pages/_app.tsx#LL11C52-L11C52

```
SwitchChainError: An error occurred when attempting to switch chain.

Details: Chain 'eip155:137' not approved. Please use one of the following: eip155:1
```

The screenshot is from `Rainbow Wallet`. As you can see in the screenshot, the networks dropdown only has Ethereum. Its should have multiple networks as supplied in the chains array. 
I tried using https://react-app.walletconnect.com/ and it shows multiple networks when trying to connect

With the walletconnect react app it looks like this, It has multiple chains in the network during connection
![Screenshot 2023-05-29 at 8 55 04 PM](https://github.com/WalletConnect/walletconnect-monorepo/assets/20742197/165ca6e5-271e-4c9d-a274-027a70ea5595)

**Reply by glitch-txs** [reply]
I was able to reproduce the error on my code as well, seems like there might be an issue with the Wagmi connector or the wallet, since it works on Trust Wallet, but at the same time Rainbow works on WalletConnect example

**Reply by glitch-txs** [reply, marked as answer]
I think I found the issue, I opened an issue on [Rainbow's repository](https://github.com/rainbow-me/rainbow/issues/4872)

**Reply by nischitpra** [reply]
Thanks alot :raised_hands:
